Azure/cloud-init.txt

```diff
@@ -0,0 +1,165 @@
+]
+}
+EOF
+
+# Load devicemapper changes
+systemctl daemon-reload
+
+# Update kernel
+echo deb http://archive.ubuntu.com/ubuntu/ xenial-proposed restricted main multiverse universe >> /etc/apt/sources.list
+apt-get update
+apt-get install -y linux-azure
+
+# Install Kubernetes and kubeadm
+apt-get install -y ebtables ethtool
+curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+cat << EOF > /etc/apt/sources.list.d/kubernetes.list
+deb http://apt.kubernetes.io/ kubernetes-xenial main
+EOF
+apt-get update
+sudo -E apt install -y kubelet=1.7.8-00 kubeadm=1.7.8-00 kubectl=1.7.8-00
+
+# Create kubelet systemd service
+cat << EOF | sudo tee /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+[Service]
+Environment="KUBELET_KUBECONFIG_ARGS=--kubeconfig=/etc/kubernetes/kubelet.conf --require-kubeconfig=true"
+Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
+Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
+Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
+Environment="KUBELET_AUTHZ_ARGS=--authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt"
+Environment="KUBELET_EXTRA_ARGS=--container-runtime=remote --container-runtime-endpoint=/var/run/crio.sock --runtime-request-timeout=30m"
+Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=0"
+ExecStart=
+ExecStart=/usr/bin/kubelet \$KUBELET_KUBECONFIG_ARGS \$KUBELET_SYSTEM_PODS_ARGS \$KUBELET_NETWORK_ARGS \$KUBELET_DNS_ARGS \$KUBELET_AUTHZ_ARGS \$KUBELET_CADVISOR_ARGS \$KUBELET_EXTRA_ARGS
+EOF
+
+# Clone, build and install CRI-O
+add-apt-repository -y ppa:projectatomic/ppa
+add-apt-repository -y ppa:alexlarsson/flatpak
+add-apt-repository -y ppa:longsleep/golang-backports
+apt-get update
+apt-get install -y btrfs-tools git golang-go libassuan-dev libdevmapper-dev libglib2.0-dev libc6-dev libgpgme11-dev libgpg-error-dev libseccomp-dev libselinux1-dev pkg-config cri-o-runc runc skopeo libostree-dev skopeo-containers go-md2man make
+
+git clone https://github.com/kubernetes-incubator/cri-o
+pushd cri-o
+git reset --hard v1.0.0
+make
+make install
+make install.config
+popd
+
+# Add CRI-O systemd service
+cat << EOF | sudo tee /etc/systemd/system/crio.service
+[Unit]
+Description=OCI-based implementation of Kubernetes Container Runtime Interface
+Documentation=https://github.com/kubernetes-incubator/cri-o
+
+[Service]
+ExecStart=/usr/local/bin/crio -log-level debug
+Restart=on-failure
+RestartSec=5
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+# Configure CRI-O
+sed -i 's/storage_driver = ""/storage_driver = "devicemapper"/' "/etc/crio/crio.conf"
+sed -i 's/storage_option = \[/storage_option = \["dm.directlvm_device=\/dev\/sdd", "dm.thinp_percent=95", "dm.thinp_metapercent=1", "dm.thinp_autoextend_threshold=80", "dm.thinp_autoextend_percent=20", "dm.directlvm_device_force=true"/' "/etc/crio/crio.conf"
```

Comment:
So it seems that devicemapper does not work with CRI-O, at least on Azure.
Our agent times out waiting for a block device to be hotplugged and thus never mount the container rootfs.

Using overlay works fine though, so those 2 lines should be replaced with one single line:

```
sed -i 's/storage_driver = ""/storage_driver = "overlay"/' "/etc/crio/crio.conf"
```

And we'll get back to devicemapper once we figure out what's happening.

Reply:
FYI, I have submitted a PR https://github.com/clearcontainers/tests/pull/634 to see how our CI behaves with devicemapper for CRIO bats.